File: bank-ocr/csharp/src/BankOcr/Parser.cs
namespace BankOcr;

public static class Parser
{
    // Canonical glyphs: each entry is the 9-char flattened 3x3 grid (top | middle | bottom).
    private static readonly IReadOnlyDictionary<string, int> Glyphs = new Dictionary<string, int>
    {
        [" _ " + "| |" + "|_|"] = 0,
        ["   " + "  |" + "  |"] = 1,
        [" _ " + " _|" + "|_ "] = 2,
        [" _ " + " _|" + " _|"] = 3,
        ["   " + "|_|" + "  |"] = 4,
        [" _ " + "|_ " + " _|"] = 5,
        [" _ " + "|_ " + "|_|"] = 6,
        [" _ " + "  |" + "  |"] = 7,
        [" _ " + "|_|" + "|_|"] = 8,
        [" _ " + "|_|" + " _|"] = 9,
    };

    public static Digit ParseDigit(IReadOnlyList<string> threeRowBlock)
    {
        if (threeRowBlock.Count != OcrDimensions.DigitHeight)
            throw new InvalidAccountNumberFormatException(
                $"Digit block must have exactly {OcrDimensions.DigitHeight} rows.");
        foreach (var row in threeRowBlock)
        {
            if (row.Length != OcrDimensions.DigitWidth)
                throw new InvalidAccountNumberFormatException(
                    $"Digit block rows must be exactly {OcrDimensions.DigitWidth} characters wide.");
        }
        var key = threeRowBlock[0] + threeRowBlock[1] + threeRowBlock[2];
        return Glyphs.TryGetValue(key, out var value) ? Digit.Of(value) : Digit.Unknown;
    }

    public static AccountNumber ParseAccountNumber(IReadOnlyList<string> threeRows)
    {
        if (threeRows.Count != OcrDimensions.DigitHeight)
            throw new InvalidAccountNumberFormatException(
                $"OCR block must have exactly {OcrDimensions.DigitHeight} rows.");
        foreach (var row in threeRows)
        {
            if (row.Length != OcrDimensions.RowWidth)
                throw new InvalidAccountNumberFormatException(
                    $"OCR block rows must be exactly {OcrDimensions.RowWidth} characters wide.");
        }

        var digits = new List<Digit>(OcrDimensions.AccountLength);
        for (var i = 0; i < OcrDimensions.AccountLength; i++)
        {
            var start = i * OcrDimensions.DigitWidth;
            var block = new[]
            {
                threeRows[0].Substring(start, OcrDimensions.DigitWidth),
                threeRows[1].Substring(start, OcrDimensions.DigitWidth),
                threeRows[2].Substring(start, OcrDimensions.DigitWidth),
            };
            digits.Add(ParseDigit(block));
        }
        return new AccountNumber(digits);
    }
}

File: bank-ocr/typescript/src/Parser.ts
import { AccountNumber } from './AccountNumber.js';
import { AccountLength, DigitHeight, DigitWidth, RowWidth } from './constants.js';
import { Digit } from './Digit.js';
import { InvalidAccountNumberFormatException } from './InvalidAccountNumberFormatException.js';

// Canonical glyphs: each entry is the 9-char flattened 3x3 grid (top | middle | bottom).
const Glyphs: ReadonlyMap<string, number> = new Map<string, number>([
  [' _ ' + '| |' + '|_|', 0],
  ['   ' + '  |' + '  |', 1],
  [' _ ' + ' _|' + '|_ ', 2],
  [' _ ' + ' _|' + ' _|', 3],
  ['   ' + '|_|' + '  |', 4],
  [' _ ' + '|_ ' + ' _|', 5],
  [' _ ' + '|_ ' + '|_|', 6],
  [' _ ' + '  |' + '  |', 7],
  [' _ ' + '|_|' + '|_|', 8],
  [' _ ' + '|_|' + ' _|', 9],
]);

export function parseDigit(threeRowBlock: readonly string[]): Digit {
  if (threeRowBlock.length !== DigitHeight) {
    throw new InvalidAccountNumberFormatException(
      `Digit block must have exactly ${DigitHeight} rows.`,
    );
  }
  for (const row of threeRowBlock) {
    if (row.length !== DigitWidth) {
      throw new InvalidAccountNumberFormatException(
        `Digit block rows must be exactly ${DigitWidth} characters wide.`,
      );
    }
  }
  const key = threeRowBlock[0]! + threeRowBlock[1]! + threeRowBlock[2]!;
  const value = Glyphs.get(key);
  return value === undefined ? Digit.unknown : Digit.of(value);
}

export function parseAccountNumber(threeRows: readonly string[]): AccountNumber {
  if (threeRows.length !== DigitHeight) {
    throw new InvalidAccountNumberFormatException(
      `OCR block must have exactly ${DigitHeight} rows.`,
    );
  }
  for (const row of threeRows) {
    if (row.length !== RowWidth) {
      throw new InvalidAccountNumberFormatException(
        `OCR block rows must be exactly ${RowWidth} characters wide.`,
      );
    }
  }
  const digits: Digit[] = [];
  for (let i = 0; i < AccountLength; i++) {
    const start = i * DigitWidth;
    const block = [
      threeRows[0]!.substring(start, start + DigitWidth),
      threeRows[1]!.substring(start, start + DigitWidth),
      threeRows[2]!.substring(start, start + DigitWidth),
    ];
    digits.push(parseDigit(block));
  }
  return new AccountNumber(digits);
}
